demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -37,72 +43,200 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
 
-    LogInfo( ( "AzureIoTPlatform_VerifyImage()\r\n" ) );
+    if( xResult != eAzureIoTSuccess )
+    {
+        AZLogError( ( "Unable to decode base64 SHA256\r\n" ) );
+        return eAzureIoTErrorFailed;
+    }
 
-    return eAzureIoTSuccess;
-}
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
 
-AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
-{
-    (void)pxAduImage;
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+
+    AZLogInfo( ( "Starting the mbedtls calculation: image size %d\r\n", pxAduImage->ulImageFileSize ) );
+
+    for( size_t ulOffset = 0; ulOffset < pxAduImage->ulImageFileSize; ulOffset += sizeof( ucPartitionReadBuffer ) )
+    {
+        ulReadSize = pxAduImage->ulImageFileSize - ulOffset < sizeof( ucPartitionReadBuffer ) ? pxAduImage->ulImageFileSize - ulOffset : sizeof( ucPartitionReadBuffer );
 
-    // TODO: Fill in to program board to decide which memory bank to use on reboot.
+        memcpy(ucPartitionReadBuffer, (pxAduImage->xUpdatePartition + ulOffset), ulReadSize);
 
-    LogInfo( ( "AzureIoTPlatform_EnableImage()\r\n" ) );
+        mbedtls_md_update( &ctx, ( const unsigned char * ) ucPartitionReadBuffer, ulReadSize );
 
+        if( ( ulOffset % 65536 == 0 ) && ( ulOffset != 0 ) )
+        {
+            printf( "." );
+        }
+    }
+
+    printf( "\r\n" );
+
+    AZLogInfo( ( "Done\r\n" ) );
+
+    mbedtls_md_finish( &ctx, ucCalculatedHash );
+    mbedtls_md_free( &ctx );
+
+    if( memcmp( ucDecodedManifestHash, ucCalculatedHash, azureiotflashSHA_256_SIZE ) == 0 )
+    {
+        AZLogInfo( ( "SHAs match\r\n" ) );
+        xResult = eAzureIoTSuccess;
+    }
+    else
+    {
+        AZLogInfo( ( "SHAs do not match\r\n" ) );
+        AZLogInfo( ( "Wanted: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            printf( "%x", ucDecodedManifestHash[ i ] );
+        }
+
+        printf( ( "\r\n" ) );
+        AZLogInfo( ( "Calculated: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            printf( "%x", ucCalculatedHash[ i ] );
+        }
+
+        printf( ( "\r\n" ) );
+
+        xResult = eAzureIoTErrorFailed;
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
+{
+    FLASH_OBProgramInitTypeDef  xOptionBytes;
+
+    HAL_FLASH_Unlock();
+    HAL_FLASH_OB_Unlock();
+    HAL_FLASHEx_OBGetConfig(&xOptionBytes); // Get current optionbytes configuration
+    xOptionBytes.OptionType = OPTIONBYTE_USER;
+    xOptionBytes.USERType = OB_USER_BFB2;
+    // If BFB2 set, we will reset it
+    xOptionBytes.USERConfig =
+        ((xOptionBytes.USERConfig & OB_BFB2_ENABLE) == OB_BFB2_ENABLE)
+            ? OB_BFB2_DISABLE
+            : OB_BFB2_ENABLE;
```

Comment:
just out of curiosity, if this check is with a bitwise, should set also be bitwise instead of `=`

Reply:
Hmm, that's interesting. This is how Azure RTOS does it as well, but I'll see if I can find some more official documentation to back this up

Reply:
Confirmed with sample from ST https://github.com/STMicroelectronics/STM32CubeL4/blob/master/Projects/32L476GDISCOVERY/Examples/FLASH/FLASH_DualBoot/Src/main.c#L114